sites/website/src/docs/design-systems/creating-a-component-library.md

```diff
@@ -0,0 +1,117 @@
+import { ElementDefinitionContext, FoundationElementDefinition } from "@microsoft/fast-foundation";
+
+const myElementStyles = (
+context: ElementDefinitionContext,
+definition: FoundationElementDefinition)
+ => css`/* ... */`;
+const myElementTemplate = (
+context: ElementDefinitionContext,
+definition: FoundationElementDefinition
+) => html`<!-- ... -->`;
+```
+
+Let’s take a closer look at what these two arguments give us.
+
+#### What is the `ElementDefinitionContext`?
+// TODO Rob - I'm not sure how to explain this argument well or what all of it's capabilities are
+
+Importantly, the ElementDefinitionContext can be used to inspect the tag-name for other components using the `ElementDefinitionContext.tagFor()` method. This allows usage of other library components in a template or stylesheet without knowing ahead of time what the tag-name for the element will be:
+
+```ts
+import { html } from "@microsoft/fast-element";
+import { FASTButton } from "@microsoft/fast-components";
+import type { ElementDefinitionContext } from "@microsoft/fast-foundation";
+
+const template = (context: ElementDefinitionContext) => {
+	const buttonTag = context.tagFor(FASTButton);
+
+	return html`<${buttonTag}>Click me!</${buttonTag}>`
+}
+```
+
+#### What is the `FoundationElementDefinition`?
+The `FoundationElementDefinition` is the configuration for which the element was defined, allowing inspection of the aggregated options configuration for the component.
+
+This definition can be extended arbitrarily by components during registration, which is what facilitates default slotted content overrides and arbitrary element configuration:
+
+```ts
+import { html } from "@microsoft/fast-element";
+import { ElementDefinitionContext, FoundationElementDefinition } from "@microsoft/fast-foundation";
+
+interface MyElementDefinition extends FoundationElementDefinition {
+    defaultContent?: string;
+}
+
+const template = (context: ElementDefinitionContext, definition: MyElementDefinition) => {
+    return html`<slot>${definition.defaultContent}</slot>`
+}
+```
+
+During component registration with the Design System, an application author can provide a `defaultContent` field as a configuration, which is used by the template if it exists.
+
+### Compose and Export Registration
+Lastly, create and export the registration function so that application authors can register the component in their Design System:
+
+```ts
+export const myElementRegistration = MyElement.compose({
+  baseName: ‘my-element’,
+  template,
+  styles,
+  defaultContent: "Click me!"
+})
+```
+
+Note that the registration is composed with a "defaultContent" value. This value can be provided during composition or as a conditional in the template; either approach will work.
+
+### Registering Library Components in an Application
+To register the component, an application author will import the registration and register it in their DesignSystem, overriding any properties as necessary:
```

Comment:
I'd like to raise a big concern here

to me a DesignSystem is smth that a Library Dev creates, while an App Dev imports it
this technically means that App Dev should not do `DesignSystem.getOrCreate()`
but this remark is less about the docs here, and more about the APIs themselves
I expected a library package to be able to export an instance of `DesignSystem` and still be able to connect this instance to a node, translating to smth like

```js
// library side
export const designSystem = new DesignSystem()
  .register(/* register all components here with all defaults */)
  .withPrefix('my');

// app side
import { designSystem } from 'my-library';
designSystem.provideFor(myAppRoot);
```

I realise that this is not how all organisation do it, but it's one of the common approaches I saw in practice and I don't understand how to best achieve it with FAST right now

Reply:
one of the ideas of how it can be done with current APIs is to export a function from the lib like this (and this is where my comment comes closer to docs which can give some guidance here):

```js
// library side
export const provideDesignSystem = (rootElement) => {
  DesignSystem
    .getOrCreate(rootElement)
    .register(/* register all components here with all defaults */).withPrefix('my');
};

// app side
import { provideDesignSystemFor } from 'my-library';
provideDesignSystem(myAppRoot);
// or for the entire document
provideDesignSystem();
```

Reply:
I think of the library dev as "defining" the design system and the app dev as "instantiating" it. The reason we have the `.getOrCreate` API is that we need to ensure that only one design system exists for any given Node in the DOM. So, we "get or create" the design system instance associated with the provided DOM Node. If none is provided, then it defaults to the document body. Would it be meaningful to add this explanation as a note in the docs?

Reply:
looks like it's not only about terminology here, but about different layers and what is expected to be done on each of them

for example `.withPrefix('my')` is not smth I want App Dev to do, but currently it's part of the "instantiating" and as such it's suggested to App Devs

I'm also starting to think that a real-world example of `getOrCreate(otherElement)` should help understand this API better
all examples in this PR (hope I didn't miss anything) are not touching upon why I would like to scope my Design System instance, and this is much related to my earlier question in Discord about why Design Tokens have their own scope (`token.setValueFor(otherElement)`) and how they are related to a Design System and it's scope

Reply:
The `provideDesignSystem` function you show is I think a good pattern. I had intended us to ship a function like this for FAST and Fluent that baked in the default prefix at least. I think we may have missed that in the release. We should patch that up.

In any case, the library dev should provide a function like that which sets up the minimal configuration for the intended design system. One of the issues with calling `register` in this case though, is that that causes a static linking of all the components, which will prevent tree shaking. The intent is that the app dev picks which components they want to register based on what they are using in their particular app. So, it should look something like this:

```ts
provideFASTDesignSystem()
  .register(
    fastButtoin()
  );
```

I'm not sure if "provide" is the best word or not. Let's continue discussion here and figure that out, then we can follow up with a PR to both FAST and Fluent to add these functions, so that at least the prefixes are set to defaults properly. That will also give us a place to do any other minimal configuration setup in the future without breaking folks by forcing them to add new configuration in successive incremental releases.

Reply:
/cc @chrisdholt This would be a non-breaking change. We'd add some functions to each library. The old mechanism would still work but we'd update docs to show this API instead as a best practice.

Reply:
> "provide" is the best word or not

I was thinking of `initDesignSystem` or just `getOrCreateDesignSystem`, but decided to use `provideDesignSystem` since it was used in the old wrapper custom element `<fast-design-system-provider>`

the old `<fast-design-system-provider>` was also responsible for design tokens, so it contained everything AFAICT, and `provide` term was a perfect match for what it was doing, whereas I don't fully understand all the implications of calling `DesignSystem.getOrCreate`, so find it difficult to suggest a good name

Reply:
I think that's a decent justification for using the term "provide" here. I'm warming up to it. Let's see if anyone else has thoughts before we commit.

Reply:
> /cc @chrisdholt This would be a non-breaking change. We'd add some functions to each library. The old mechanism would still work but we'd update docs to show this API instead as a best practice.

@EisenbergEffect I have a use case I'm curious about with the above. Lets say I'm a library author extending the FAST or Fluent design system with my own components, but I *want* my own prefix - IE, I don't want to prefix my components with Fluent, but I do want to retain that prefix for the Fluent components. Is there a mechanism we can have in `provideFluentDesignSystem`, where I register my own components under the same design system bug with a custom prefix?

Reply:
Yes. It would look like this:

```ts
provideFluentDesignSystem(someChildElement)
  .register(
    fluentButton()
  ).withPrefix('min')
  .register(
    myComponent()
  );
```

Reply:
Or, you could do it like this:

```ts
provideFluentDesignSystem()
  .register(
    fluentButton()
  );

DesignSystem.getOrCreate()
  .withPrefix('mine')
  .register(
    myComponent()
  );
```

Reply:
I'm thinking of

```ts
// library
export const provideMyDesignSystem = (childElement?: HTMLElement) => {
  const ds = DesignSystem.getOrCreate(childElement).withPrefix('my');

  // ... setup whatever other defaults are necessary ...

  return {
    register: ds.register,
  };
};

// export all components registration functions
export { myAccordion, myButton }; 
```

if one needs to have full control, there is always a possibility to import `DesignSystem` from '@microsoft/fast-foundation' directly and setup everything from scratch, as individual components are necessary to export for tree-shaking anyway according to an earlier remark

Reply:
That could absolutely be the pattern that you follow with your library, and we could provide guidance on that. I don't think we'd want to only export `register` from our own component libraries (FAST and Fluent), as that would cut off some scenarios we know people need that extra configuration for. But, I do think this makes sense.

Reply:
Fully agree!

I came to the following layers when analysing a bigger picture and see I overlooked 2 of them which give more sense to certain APIs:

1. foundation components layer: maximum flexibility, sensible defaults
2. company components layer: use the flexibility of the foundation layer, but enforce certain things (e.g. prefix) and close them for modification, export individual component for app to decide what to register
3. app setup layer: register only the components needed by the app
4. feature level: only use the components available in the app

What is restricted and allowed on layers 2-4 is up to the company building their Design System, but the foundation layer of components you are building in FAST, including `@microsoft/fast-components` should give this flerxibility

This brings me to the next question/suggestion: although be default almost everything should be customisable on your 1st layer, there should be ways to restrict customisation of anything on layers 2-3, while preserving customisation for what is needed to remain customisable. Docs might help to understand this better, but maybe some new APIs will be needed as well.

PS: I'm now translating my vision into a Design System starter kit and will be able to tell more once I make more progress with it. E.g. if some APIs are missing or if smth is hard to achieve.

Reply:
I'm going to work to get the `provideFASTDesignSystem()` API in soon. Then I'll update the docs to use that throughout. I like the content of you've got here on layers. So, I'll also add a guidance section to the creating libraries doc and include some of that along with the other things we've been talking about. Thanks for all the time you've taken to provide this feedback @bashmish

Reply:
@EisenbergEffect thanks for constructive communication about it! Feel free to move some of my remaining remarks into future PRs, maybe this PR is good enough for now to update docs on the FAST website and help people onboard with version 2.x.x.

Reply:
I've captured these notes so that I can follow up with a PR to improve the ergonomics and accompany that with docs updates. Going to resolve this feedback for now.